include/ensmallen_bits/nsga2/nsga2.hpp

```diff
@@ -89,12 +89,12 @@ class NSGA2 {
    * @param objectives Vector of objective functions to optimize for.
    * @param iterate Starting point.
    * @param callbacks Callback functions.
-   * @return std::vector<MatType> The best front obtained after optimization.
+   * @return std::vector<arma::Col<double>> Objective values for each of the children.
    */
   template<typename MatType,
            typename... ArbitraryFunctionType,
            typename... CallbackTypes>
-  std::vector<MatType> Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
+  std::vector<arma::Col<double>> Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
```

Comment:
Hm, I was thinking we could take the same over the objectives, in which case we should also provide a method `Objectives` to return the objectives in a vector.

Reply:
I don't think I understand your comment. Erm, Marcus could you please explain it further. 😅

Reply:
I was referring to the loss, the idea is to also return a single double, does that make sense?

Reply:
Ah, uhm since we have multi objectives, we have to reduce the individual values from each objective, right? Also, I don't know of any indicator that can summarise the 'fitness' of the entire population (like the loss does in an ANN). 

What do you suggest?

Reply:
I think that to match the API of the rest of the optimizers, it could be just fine to return a sum of the objectives, and then also perhaps provide an overload like

```
typename MatType::elem_type Optimize(
    std::tuple<ArbitraryFunctionType...>& objectives,
    MatType& iterate,
    arma::Col<typename MatType::elem_type>& finalObjectives,
    CallbackTypes... callbacks);
```

and in this case `finalObjectives` will be filled with the final objective value.  What do you think?

Reply:
Ah, I get it now. Sure, I'm looking into it.

Reply:
@rcurtin just to be sure, `calculatedObjectives` is of type `std::vector<arma::Col<MatType::elem_type>>` and we are reduces them by addition across the vector, right? 

But, this will yield a `arma::Col<MatType::elem_type>` instead of the `MatType::elem_type` you suggested.

How do we get past this?

Reply:
I'm not sure why the original return type you had was `std::vector<arma::Col<double>>`.  If that is supposed to be the objective value at the final iterate position for each individual objective function, shouldn't that just be `arma::Col<typename MatType::elem_type>`?  In any case, if you do have an `arma::Col` that contains all of the individual objective values, then you can just return `accu()` over that column vector.  Let me know if I missed something.

Reply:
I was returning `std::vector<arma::Col<double>>` as it contained the evaluated objectives for all children across all objectives. So the vector was for each child in the population and then the column vector was for the multiple objectives.

Reply:
When we call `Optimize()`, the user expects that "the best" coordinates are stored in the `MatType& iterate` parameter.  So in this sense the `Optimize()` function is a black-box interface in which you put a set of objective functions and get back one "best" coordinate.  So we should not return any information about all of the children in the population, just the best result.  Does that make sense?  Did I overlook anything?

Reply:
Erm, in NSGA2, the algorithm outputs a set of "best" coordinates, not a single coordinate. Will I store a random sample from this set in `MatType& iterate`?

Reply:
The first one is reasonable to me, we should still provide a method to access all, but that's easy to add.

Reply:
Btw. thanks for keeping up with all the comments.

Reply:
Marcus, the entire set of coordinates in the Pareto front are accessible using `Front()`.
Also, it's been a pleasure working with all of you! 😄

Reply:
Ahh, right, I see what you mean.  It should definitely be possible to recover all the coordinates on the Pareto frontier (by the way should we call that `Frontier()`?  I'm not sure what the more commonly used terminology is.).

For each set of coordinates on the Pareto front, there are an associated set of objective values with each objective function.  So your choice to provide a `std::vector<arma::Col<typename MatType::elem_type>>` now makes sense to me---the `std::vector<>` wraps each set of coordinates in the Pareto frontier, and then the `arma::Col<>` holds the objective of every individual objective function associated with that set of coordinates.

So you can see that we can take the sum of each of those `arma::Col<typename MatType::elem_type>`s in order to produce a single `typename MatType::elem_type`, whose value is just the sum of all objectives for a given set of coordinates.  But if we want to return a single element representing how good the optimization was from `Optimize()`, then perhaps it makes sense to take the minimum of that summation across all sets of coordinates?  I'm not sure what the best option here.  It seems like that quantity could be useful as an indicator of how successful the optimization was.

Overall, my hope is to maximize the similarity to all of the other `Optimize()` functions from all other optimizer types; if I overlooked something or if my suggestion doesn't make sense from the perspective of the multi-objective optimization literature, then we don't have to go with it. :smile:

Reply:
> (by the way should we call that Frontier()? I'm not sure what the more commonly used terminology is.).

All these terms are used interchangeably. Should we provide aliases?

> But if we want to return a single element representing how good the optimization was from Optimize(), then perhaps it makes sense to take the minimum of that summation across all sets of coordinates?

Sure, this makes sense. 😃